bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogManagerForEntryLogPerLedger.java

```diff
@@ -0,0 +1,463 @@
+            @Override
+            public void diskFull(File disk) {
+                Set<BufferedLogChannel> copyOfCurrentLogs = getCopyOfCurrentLogs();
+                for (BufferedLogChannel currentLog : copyOfCurrentLogs) {
+                    if (disk.equals(currentLog.getLogFile().getParentFile())) {
+                        currentLog.setLedgerDirFull(true);
+                    }
+                }
+            }
+
+            @Override
+            public void diskWritable(File disk) {
+                Set<BufferedLogChannel> copyOfCurrentLogs = getCopyOfCurrentLogs();
+                for (BufferedLogChannel currentLog : copyOfCurrentLogs) {
+                    if (disk.equals(currentLog.getLogFile().getParentFile())) {
+                        currentLog.setLedgerDirFull(false);
+                    }
+                }
+            }
+        };
+    }
+
+    public void acquireLock(Long ledgerId) throws IOException {
+        try {
+            ledgerIdEntryLogMap.get(ledgerId).getLedgerLock().lock();
+        } catch (Exception e) {
+            log.error("Received unexpected exception while fetching lock to acquire", e);
+            throw new IOException("Received unexpected exception while fetching lock to acquire", e);
+        }
+    }
+
+    public void releaseLock(Long ledgerId) throws IOException {
+        try {
+            ledgerIdEntryLogMap.get(ledgerId).getLedgerLock().unlock();
+        } catch (Exception e) {
+            log.error("Received unexpected exception while fetching lock to release", e);
+            throw new IOException("Received unexpected exception while fetching lock to release", e);
+        }
+    }
+
+    /*
+     * sets the logChannel for the given ledgerId. It will add the new
+     * logchannel to replicaOfCurrentLogChannels, and the previous one will
+     * be removed from replicaOfCurrentLogChannels. Previous logChannel will
+     * be added to rotatedLogChannels in both the cases.
+     */
+    @Override
+    public void setCurrentLogForLedgerAndAddToRotate(long ledgerId, BufferedLogChannel logChannel) throws IOException {
+        acquireLock(ledgerId);
+        try {
+            BufferedLogChannel hasToRotateLogChannel = getCurrentLogForLedger(ledgerId);
+            logChannel.setLedgerIdAssigned(ledgerId);
+            ledgerIdEntryLogMap.get(ledgerId).setEntryLog(logChannel);
+            replicaOfCurrentLogChannels.put(logChannel.getLogId(), logChannel);
+            if (hasToRotateLogChannel != null) {
+                replicaOfCurrentLogChannels.remove(hasToRotateLogChannel.getLogId());
+                rotatedLogChannels.add(hasToRotateLogChannel);
+            }
+        } catch (Exception e) {
+            log.error("Received unexpected exception while fetching entry from map", e);
+            throw new IOException("Received unexpected exception while fetching entry from map", e);
+        } finally {
+            releaseLock(ledgerId);
```

Comment:
+1 on what @dlg99 pointed out

Reply:
changed the logic of releaselock in the last minute, because of some issue in getIfPresent. So now it doesn't makes complete sense to do what I'm doing.

Will make changes as Andrey pointed out.